core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala

```diff
@@ -0,0 +1,533 @@
+    val setDataRequest = SetDataRequest(TopicZNode.path(topic), TopicZNode.encode(assignment), -1, null)
+    retryUntilConnected(setDataRequest).asInstanceOf[SetDataResponse]
+  }
+
+  def getAllLogDirEventNotifications = {
+    val getChildrenResponse = retryUntilConnected(GetChildrenRequest(LogDirEventNotificationZNode.path, null)).asInstanceOf[GetChildrenResponse]
+    if (Code.get(getChildrenResponse.rc) == Code.OK) {
+      getChildrenResponse.children.map(LogDirEventNotificationSequenceZNode.sequenceNumber)
+    } else if (Code.get(getChildrenResponse.rc) == Code.NONODE) {
+      Seq.empty
+    } else {
+      throw KeeperException.create(Code.get(getChildrenResponse.rc))
+    }
+  }
+
+  def getBrokerIdsFromLogDirEvents(sequenceNumbers: Seq[String]): Seq[Int] = {
+    val getDataRequests = sequenceNumbers.map(sequenceNumber => GetDataRequest(LogDirEventNotificationSequenceZNode.path(sequenceNumber), null))
+    val getDataResponses = retryUntilConnected(getDataRequests).map(_.asInstanceOf[GetDataResponse])
+    getDataResponses.flatMap { getDataResponse =>
+      if (Code.get(getDataResponse.rc) == Code.OK) {
+        LogDirEventNotificationSequenceZNode.decode(getDataResponse.data)
+      } else if (Code.get(getDataResponse.rc) == Code.NONODE) {
+        None
+      } else {
+        throw KeeperException.create(Code.get(getDataResponse.rc))
+      }
+    }
+  }
+
+  def deleteLogDirEventNotifications(): Unit = {
+    val getChildrenResponse = retryUntilConnected(GetChildrenRequest(LogDirEventNotificationZNode.path, null)).asInstanceOf[GetChildrenResponse]
+    if (Code.get(getChildrenResponse.rc) == Code.OK) {
+      deleteLogDirEventNotifications(getChildrenResponse.children)
+    } else if (Code.get(getChildrenResponse.rc) != Code.NONODE) {
+      throw KeeperException.create(Code.get(getChildrenResponse.rc))
+    }
+  }
+  def deleteLogDirEventNotifications(sequenceNumbers: Seq[String]): Unit = {
+    val deleteRequests = sequenceNumbers.map(sequenceNumber => DeleteRequest(LogDirEventNotificationSequenceZNode.path(sequenceNumber), -1, null))
+    retryUntilConnected(deleteRequests)
+  }
+
+  def getReplicaAssignmentForTopics(topics: Set[String]): Map[TopicAndPartition, Seq[Int]] = {
+    val getDataRequests = topics.map(topic => GetDataRequest(TopicZNode.path(topic), topic))
+    val getDataResponses = retryUntilConnected(getDataRequests.toSeq).map(_.asInstanceOf[GetDataResponse])
+    getDataResponses.flatMap { getDataResponse =>
+      val topic = getDataResponse.ctx.asInstanceOf[String]
+      if (Code.get(getDataResponse.rc) == Code.OK) {
+        TopicZNode.decode(topic, getDataResponse.data)
+      } else if (Code.get(getDataResponse.rc) == Code.NONODE) {
+        Map.empty[TopicAndPartition, Seq[Int]]
+      } else {
+        throw KeeperException.create(Code.get(getDataResponse.rc))
+      }
+    }.toMap
+  }
+
+  def getTopicDeletions = {
```

Comment:
Could we explicitly define the return type of this method?

Reply:
Done.